artiq/frontend/aqctl_proxy_moninj.py

```diff
@@ -0,0 +1,322 @@
+
+    def on_injection_status(self, channel, override, value):
+        logger.debug(f"received injection status {(channel, override, value)}")
+        for client in self.proxy.clients:
+            if (channel, override) in client.injection:
+                # InjectionStatus
+                packet = struct.pack(self.comm.endian + "blbb", 1, channel, override, value)
+                client.writer.write(packet)
+
+
+class MonInjMaster:
+    def __init__(self, proxy):
+        self._core_addr = None
+        self.proxy = proxy
+        self.ddb = dict()
+        self.ddb_notify = Subscriber("devices", self.build_ddb, self.on_notify, self.on_disconnected)
+        self.connected = False
+
+    def build_ddb(self, init):
+        self.ddb = init
+        return self.ddb
+
+    async def connect(self, master, port):
+        logger.debug(f"trying to connect to master server at {master}:{port}")
+        await self.ddb_notify.connect(master, port)
+
+    async def disconnect(self):
+        await self.ddb_notify.close()
+        await self.on_disconnected()
+
+    @property
+    def core_addr(self):
+        return self._core_addr
+
+    async def set_core_addr(self, value):
+        if self._core_addr and value and self._core_addr != value:
+            logging.debug(f"core address changed, old: {self._core_addr}, new: {value}")
+            await self.proxy.core.disconnect()
+        self._core_addr = value
+
+    # Callbacks
+    async def on_connected(self):
+        if not self.connected:
+            self.connected = True
+            logger.info("connected to master")
+            await self.set_core_addr(self.ddb["core"]["arguments"]["host"])
+            if not self.proxy.core.connected:
+                await self.proxy.core.connect(self.core_addr, 1383)
+
+    async def on_disconnected(self):
+        if self.connected:
+            self.connected = False
+            logger.info("disconnected from master")
+            while self.proxy.active and not self.connected:
+                try:
+                    await self.connect(self.proxy.master_addr, self.proxy.master_notify_port)
+                except:
+                    logging.error("master still unreachable, retrying...")
+                    await asyncio.sleep(10)
+
+    async def on_notify(self, mod):
+        logger.debug(f"received mod from master {mod}")
+        if mod["action"] == "init":
+            await self.on_connected()
+        if mod["action"] == "setitem" and mod["key"] == "core":
+            await self.set_core_addr(mod["value"]["arguments"]["host"])
+
+
+class MonInjProxy(AsyncioServer):
+    def __init__(self, master, notify_port):
+        super().__init__()
+        self.master_addr = master
+        self.master_notify_port = notify_port
+        self.core, self.master = MonInjCoredev(self), MonInjMaster(self)
+        self.active = False
+        self.mon_fields = MonitoredField()
+        self.clients = []
+        self.endian = ""
```

Comment:
AFAICT (barely) you just funnel the endianness from comm in `on_connected`.
Not sure if we can just directly access it through `self.core.comm.endian` in `_handle_connection_cr` instead. Though this looks a bit messy as well.

Reply:
done